### ISSUE TITLE
Rule proposal: Array#push + rest parameter instead of Arrray#concat

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1250]
Just a minor suggestion, not sure if it's actually any better or preferred. Maybe:

- `array` can be a `const`
- no repetition
- shorter

## Fail

```js
array = array.concat(otherArray);
```

## Pass

```js
array.push(...otherArray);
```

**Comment by sindresorhus** [comment]
We already fix `array = array.concat(otherArray)` to `array = [...array, ...otherArray]` in the `prefer-spread` rule.

**Comment by fregante** [comment]
I suppose my “real-life example” wasn’t caught because it was incorrect. But I still think the .push version is better in every way (shorter, no repetition, less looping via spread, no creation of another array in memory)

**Comment by fregante** [comment]
I suppose the difference would be that with `.push`, `array` would be modified, so if it's a parameter function it's not a safe fix:

```js
function spread(array) {
	const otherArray = [1, 2, 3];
	array = [...array, ...otherArray];
	return array;
}
const initial = [0];
const newOne = spread(initial);

// initial => [0]
// newOne => [0, 1, 2, 3]
```


```js
function push(array) {
	const otherArray = [1, 2, 3];
	array.push(…otherArray);
	return array;
}
const initial = [0];
const newOne = append(initial);

// initial => [0, 1, 2, 3]
// newOne => [0, 1, 2, 3]
```

**Comment by papb** [comment]
I've heard that JavaScript can pass at most 30 thousand arguments to a function, could `.push(...foo)` (or any other function call with the spread operator) hit this limitation as well if the spreaded array has length above 30000?

**Comment by fisker** [comment]
```js
[].push(...Array.from({length: 3e6}))

Uncaught RangeError: Maximum call stack size exceeded
    at <anonymous>:1:4
```


```js
[...Array.from({length: 3e6})] // can run
```

**Comment by fisker** [comment]
Check this https://github.com/prettier/prettier/pull/7587#discussion_r379907099

**Comment by fisker** [comment]
https://dev.to/uilicious/javascript-array-push-is-945x-faster-than-array-concat-1oki

**Comment by fregante** [comment]
> I've heard that JavaScript can pass at most 30 thousand arguments to a function

Could this be part of the error message, without an autofix?

```
Prefer array.push to .concat if the array is smaller than 5000 items // for safety
```

**Comment by sindresorhus** [comment]
I don't think we should do this

The safe part is already covered: `prefer-spread` turns `array = array.concat(otherArray)` into `array = [...array, ...otherArray]`, preserving `concat`'s non-mutating semantics. The actual bug case from the linked ESLint issue, discarded `array.concat(otherArray)`, is also covered now by `no-unused-array-method-return`

`array.push(...otherArray)` is a different contract. It mutates the receiver, which changes behvior for aliases, parameters, and shared arrays, and it can fail for large arrays because spread in a call hits engine argument limits. That makes this a performance/style choice for specific hot paths, not a safe general lint preference.
